core/src/main/java/org/infinispan/interceptors/impl/PrefetchInterceptor.java

```diff
@@ -207,68 +214,96 @@ private Object prefetchKeysIfNeededAndInvokeNext(InvocationContext ctx, Visitabl
-         }));
+         PutKeyValueCommand putKeyValueCommand = commandsFactory.buildPutKeyValueCommand(cmd.getKey(), maxValue.getValue(), maxValue.getMetadata(), STATE_TRANSFER_FLAGS);
+         putKeyValueCommand.setTopologyId(cmd.getTopologyId());
+         invokeNext(ctx1, putKeyValueCommand);
       }));
    }
 
-   private CompletableFuture<InternalCacheValue> retrieveRemoteValue(Collection<Address> targets, Object key) {
+   private CompletableFuture<InternalCacheValue> retrieveRemoteValue(Collection<Address> targets, Object key, int topologyId) {
+      if (trace) {
+         log.tracef("Prefetching entry for key %s from %s", key, targets);
+      }
       ClusteredGetCommand command = commandsFactory.buildClusteredGetCommand(key, FlagBitSets.SKIP_OWNERSHIP_CHECK);
-      return rpcManager.invokeRemotelyAsync(targets, command, syncRpcOptions).thenApply(responseMap -> {
-         EntryVersion maxVersion = null;
-         InternalCacheValue maxValue = null;
-         for (Response response : responseMap.values()) {
-            if (!response.isSuccessful()) {
-               throw OutdatedTopologyException.INSTANCE;
-            }
-            SuccessfulResponse successfulResponse = (SuccessfulResponse) response;
-            InternalCacheValue icv = (InternalCacheValue) successfulResponse.getResponseValue();
-            if (icv == null) {
-               continue;
-            }
-            Metadata metadata = icv.getMetadata();
-            if (metadata instanceof RemoteMetadata) {
-               // not sure if this can happen, but let's be on the safe side
-               throw OutdatedTopologyException.INSTANCE;
-            }
-            if (metadata != null && metadata.version() != null) {
-               if (maxVersion == null || maxVersion.compareTo(metadata.version()) == InequalVersionComparisonResult.BEFORE) {
-                  maxVersion = metadata.version();
-                  maxValue = icv;
-               }
+      command.setTopologyId(topologyId);
+      return rpcManager.invokeRemotelyAsync(targets, command, syncRpcOptions).thenApply(handleRemotelyPrefetchedEntry);
+   }
+
+   private InternalCacheValue handleRemotelyPrefetchedEntry(Map<Address, Response> responseMap) {
+      EntryVersion maxVersion = null;
+      InternalCacheValue maxValue = null;
+      for (Response response : responseMap.values()) {
+         if (!response.isSuccessful()) {
+            throw OutdatedTopologyException.INSTANCE;
+         }
+         SuccessfulResponse successfulResponse = (SuccessfulResponse) response;
+         InternalCacheValue icv = (InternalCacheValue) successfulResponse.getResponseValue();
+         if (icv == null) {
+            continue;
+         }
+         Metadata metadata = icv.getMetadata();
+         if (metadata instanceof RemoteMetadata) {
+            // not sure if this can happen, but let's be on the safe side
+            throw OutdatedTopologyException.INSTANCE;
```

Comment:
I think it can happen if the remote get is delayed a lot and the key ownership changes from A to B and then to A again.